runtime/src/main/kotlin/kotlin/Numbers.kt

```diff
@@ -100,4 +100,180 @@ public actual inline fun Float.Companion.fromBits(bits: Int): Float = kotlin.fro
+
+/**
+ * Counts the number of consecutive most significant bits that are zero in the binary representation of this [Int] number.
+ */
+@SinceKotlin("1.3")
+@ExperimentalStdlibApi
+public actual fun Int.countLeadingZeroBits(): Int =
+        if (this == 0) Int.SIZE_BITS else countLeadingZeroBits(this)
+
+/**
+ * Counts the number of consecutive least significant bits that are zero in the binary representation of [Int] [value].
+ * Returns undefined result for zero [value].
+ */
+@SymbolName("Kotlin_Int_countTrailingZeroBits")
+private external fun countTrailingZeroBits(value: Int): Int
+
+/**
+ * Counts the number of consecutive least significant bits that are zero in the binary representation of this [Int] number.
+ */
+@SinceKotlin("1.3")
+@ExperimentalStdlibApi
+public actual fun Int.countTrailingZeroBits(): Int =
+        if (this == 0) Int.SIZE_BITS else countTrailingZeroBits(this)
+
+/**
+ * Returns a number having a single bit set in the position of the most significant set bit of this [Int] number,
+ * or zero, if this number is zero.
+ */
+@SinceKotlin("1.3")
+@ExperimentalStdlibApi
+public actual fun Int.takeHighestOneBit(): Int =
+        if (this == 0) 0 else 1.shl(Int.SIZE_BITS - 1 - countLeadingZeroBits(this))
+
+/**
+ * Returns a number having a single bit set in the position of the least significant set bit of this [Int] number,
+ * or zero, if this number is zero.
+ */
+@SinceKotlin("1.3")
+@ExperimentalStdlibApi
+public actual fun Int.takeLowestOneBit(): Int =
+        this and -this
+
+/**
+ * Rotates the binary representation of this [Int] number left by the specified [bitCount] number of bits.
+ * The most significant bits pushed out from the left side reenter the number as the least significant bits on the right side.
+ *
+ * Rotating the number left by a negative bit count is the same as rotating it right by the negated bit count:
+ * `number.rotateLeft(-n) == number.rotateRight(n)`
+ *
+ * Rotating by a multiple of [Int.SIZE_BITS] (32) returns the same number, or more generally
+ * `number.rotateLeft(n) == number.rotateLeft(n % 32)`
+ */
+@SinceKotlin("1.3")
+@ExperimentalStdlibApi
+public actual fun Int.rotateLeft(bitCount: Int): Int =
+        shl(bitCount) or ushr(Int.SIZE_BITS - bitCount)
+
+
+/**
+ * Rotates the binary representation of this [Int] number right by the specified [bitCount] number of bits.
+ * The least significant bits pushed out from the right side reenter the number as the most significant bits on the left side.
+ *
+ * Rotating the number right by a negative bit count is the same as rotating it left by the negated bit count:
+ * `number.rotateRight(-n) == number.rotateLeft(n)`
+ *
+ * Rotating by a multiple of [Int.SIZE_BITS] (32) returns the same number, or more generally
+ * `number.rotateRight(n) == number.rotateRight(n % 32)`
+ */
+@SinceKotlin("1.3")
+@ExperimentalStdlibApi
+public actual fun Int.rotateRight(bitCount: Int): Int =
+        shl(Int.SIZE_BITS - bitCount) or ushr(bitCount)
```

Comment:
How does it handle incorrect values?

Reply:
All `bitCount` values are correct. Shift operations take five least significant bits of bitCount argument, so for example `x.shl(-1) == x.shl(31)`